include/fsm/fsm.h

```diff
@@ -254,6 +254,12 @@ fsm_mapendids(struct fsm * fsm,
 	int (*remap)(fsm_state_t state, size_t num_ids, fsm_end_id_t *endids, size_t *num_written, void *opaque),
 	void *opaque);
 
+/* Remaps endids by adding a constant delta to them.  Note that this will wrap around as an unsigned integer,
```

Comment:
> It's good that you're calling out the possibility of rollover here. If that happens, could it break internal invariants about being sorted and/or trigger asserts?

I don't believe so.  `fsm_increndids` uses `fsm_mapendids` behind the scenes, which ensures that the ids are sorted.  Looking at the code, I should also have `fsm_mapendids` check for duplicates.

> My expectation for how these endids will be used is that they will be fairly small and sequentially allocated (0 1 2 3 4...n) rather than sparse (0 10000 123456 850000 999999), so realistically rollover should only happen in pathological cases, but I don't think that's actually articulated anywhere. Does that track your mental model for how those should be used?

I didn't want to impose a model on the end ids.  I  could imagine applications where the value was a hash that was used to find a bucket, or some other sparse value.  I'm not sure how realistic these uses are, though.  I plan to use them the way you describe

> If they are expected to be small and non-sparse, using include/adt/pv.h's permutation_vector could avoid the overhead from qsort, which can be surprisingly high when lots of small arrays of numbers are being sorted.

Another approach would be to use a hash set.  I'm not sure; I'm tempted to leave things as-is unless it proves to be a problem.

Reply:
Added duplicate removal (and test for this) in [d67119f](https://github.com/katef/libfsm/pull/426/commits/d67119fced878a14ec66298504a56d63c01eb709)